sycl/include/CL/sycl/exception.hpp

```diff
@@ -12,40 +12,46 @@
 
 #include <CL/sycl/detail/common.hpp>
 #include <CL/sycl/stl.hpp>
-#include <exception>
+
+#include <cstddef>
 
 namespace cl {
 namespace sycl {
 
+// Forward declaration
 class context;
 
 struct exception {
+
+  using ContextPtr = shared_ptr_class<context>;
+
   exception() = default;
 
-  const char *what() const noexcept { return msg.c_str(); }
+  exception(const char *Msg, const cl_int ClErr = CL_SUCCESS,
+            ContextPtr Context = nullptr)
+      : Msg(std::string(Msg) + " " +
+            ((ClErr == CL_SUCCESS) ? "" : OCL_CODE_TO_STR(ClErr))),
+        ClErr(ClErr), Context(Context) {}
+
+  exception(const std::string &Msg, const cl_int ClErr = CL_SUCCESS,
+            ContextPtr Context = nullptr)
```

Comment:
Why do you need a shared pointer to a context, if you only have to return that context by value?

Reply:
None of these seems like an argument: using a shared ptr accompanied by a heap allocation is a huge overhead _just_ to get an additional state (context is missing). One extra bool variable should do the trick.

Reply:
Will be done in a separate patch. This patch I will leave the as NFC.